### PR TITLE
Query generator kind for error reporting

### PR DESCRIPTION
**Pull request by csmoe**
Fixes https://github.com/rust-lang/rust/pull/65166#discussion_r333114545
r? @estebank 
cc @cramertj

### CODE REVIEW

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 1]
```diff
@@ -1223,12 +1224,22 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             },
             Err(_) => "move |<args>| <body>".to_string()
         };
-
+        let kind = if let Some(generator_kind) = use_span.generator_kind() {
```
Comment (Centril) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: Flatten this into the `match`?

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 2]
```diff
@@ -1223,12 +1224,22 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
...
+            match generator_kind {
+                GeneratorKind::Async(_) => "async block",
```
Comment (cramertj) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: This should match on `AsyncGeneratorKind` (the field inside `Async(_)`) to differentiate async blocks, closures, and functions (though I'd expect only the former two to appear here).

File: src/librustc_mir/borrow_check/conflict_errors.rs [hunk 3]
```diff
@@ -1223,12 +1224,25 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
             },
             Err(_) => "move |<args>| <body>".to_string()
         };
-
+        let kind = match use_span.generator_kind() {
+            Some(generator_kind) => match generator_kind {
+                GeneratorKind::Async(async_kind) => match async_kind {
+                    AsyncGeneratorKind::Block
+                    | AsyncGeneratorKind::Closure => "async block",
+                    _ => bug!("async block expected, but async funtion founded."),
```
Comment (estebank) [inline src/librustc_mir/borrow_check/conflict_errors.rs]: But async function found.

File: src/librustc_mir/borrow_check/error_reporting.rs [hunk 4]
```diff
@@ -1,10 +1,11 @@
 use rustc::hir;
 use rustc::hir::def::Namespace;
 use rustc::hir::def_id::DefId;
+use rustc::hir::GeneratorKind;
 use rustc::mir::{
     AggregateKind, Constant, Field, Local, LocalKind, Location, Operand,
     Place, PlaceBase, PlaceRef, ProjectionElem, Rvalue, Statement, StatementKind,
-    Static, StaticKind, Terminator, TerminatorKind,
+    Static, StaticKind, Terminator, TerminatorKind
```
Comment (estebank) [inline src/librustc_mir/borrow_check/error_reporting.rs]: The trailing comma was ok

**Review by estebank — APPROVED** [review]

**Pull request merged**